### ISSUE TITLE
Using a custom collate_fn with testers and logging_presets

### ISSUE THREAD

**Issue by tlscabinet** [KevinMusgrave/pytorch-metric-learning#289]
Hello, it is not clear how can I use RNN architecture with this package, basically my problem is that I have to pad the sequences in each batch, but I don't have the access to the DataLoader, how should I approach this problem?

**Comment by KevinMusgrave** [comment]
Which part of this package are you trying to use (losses, miners, trainers, etc.)?

**Comment by tlscabinet** [comment]
I can pass the collocate function into trainer, but there is no argument of collocate function in tester, how can I use collocate function in tester?

**Comment by KevinMusgrave** [comment]
You can pass the collate_fn into tester.test as a keyword argument:

```python
tester.test(dataset_dict, epoch, model, collate_fn=collate_fn)
```

This isn't documented, sorry about that

**Comment by tlscabinet** [comment]
I have tried this, it doesn’t work, instead I have put the batch size equal
to 1 during the testing which have solved (partially) the problem

**Comment by KevinMusgrave** [comment]
So the collate_fn works in the trainer but not the tester?

**Comment by tlscabinet** [comment]
Yeah, exactly

**Comment by KevinMusgrave** [comment]
Hmm 🤔

**Comment by tlscabinet** [comment]
> Hmm 🤔
One more small question: during the validation I want to train my classification models, how can I do it? basically during the validation I need to have the access to training emb, labels, val emb and their labels.

**Comment by KevinMusgrave** [comment]
If you want to do that after accuracy has been computed, you can use [```end_of_testing_hook```](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/de47c68ab19ba606979221c3629f74ea729eff29/src/pytorch_metric_learning/testers/base_tester.py#L308). You can pass it in like this:
```python
# define the function
def your_end_of_testing_hook(tester):
    train_emb, train_labels = tester.embeddings_and_labels["train"]
    val_emb, val_labels = tester.embeddings_and_labels["val"]

# pass it in when you create the tester
tester = GlobalEmbeddingSpaceTester(end_of_testing_hook = your_end_of_testing_hook)
```

But if you want to train at some other point in the ```tester.test``` function, there isn't a simple way of doing it.

**Comment by KevinMusgrave** [comment]
Or you could write your own class that extends ```GlobalEmbeddingSpaceTester``` and override [get_all_embeddings_for_all_splits](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/de47c68ab19ba606979221c3629f74ea729eff29/src/pytorch_metric_learning/testers/base_tester.py#L247-L261), since that occurs before accuracy is calculated.

```python
class CustomTester(GlobalEmbeddingSpaceTester):
    def get_all_embeddings_for_all_splits(
        self,
        dataset_dict,
        trunk_model,
        embedder_model,
        splits_to_compute_embeddings,
        collate_fn,
    ):
        embeddings_and_labels = {}
        for split_name in splits_to_compute_embeddings:
            logging.info("Getting embeddings for the %s split" % split_name)
            embeddings_and_labels[split_name] = self.get_all_embeddings(
                dataset_dict[split_name], trunk_model, embedder_model, collate_fn
            )

        # do the training here        

        return embeddings_and_labels

```

**Comment by tlscabinet** [comment]
> Or you could write your own class that extends `GlobalEmbeddingSpaceTester` and override [get_all_embeddings_for_all_splits](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/de47c68ab19ba606979221c3629f74ea729eff29/src/pytorch_metric_learning/testers/base_tester.py#L247-L261), since that occurs before accuracy is calculated.
> 
> ```python
> class CustomTester(GlobalEmbeddingSpaceTester):
>     def get_all_embeddings_for_all_splits(
>         self,
>         dataset_dict,
>         trunk_model,
>         embedder_model,
>         splits_to_compute_embeddings,
>         collate_fn,
>     ):
>         embeddings_and_labels = {}
>         for split_name in splits_to_compute_embeddings:
>             logging.info("Getting embeddings for the %s split" % split_name)
>             embeddings_and_labels[split_name] = self.get_all_embeddings(
>                 dataset_dict[split_name], trunk_model, embedder_model, collate_fn
>             )
> 
>         # do the training here        
> 
>         return embeddings_and_labels
> ```

But i am already passing end_of_testing_hook = hooks.end_of_testing_hook into the tester..

**Comment by KevinMusgrave** [comment]
That's ok, you can wrap it

```python
# define the function
def your_end_of_testing_hook(tester):
    train_emb, train_labels = tester.embeddings_and_labels["train"]
    val_emb, val_labels = tester.embeddings_and_labels["val"]

    # do training here

    hooks.end_of_testing_hook(tester)
```

**Comment by tlscabinet** [comment]
Oh okay, thank you for help)

**Comment by tlscabinet** [comment]
Hello once more, I have faced some really strange issue: I am using RNN model and I am working with two strategies 1) I am padding the data in advance and saving my features afterwards I am using pytorch metric learning package to learn the embedding space 2) I am doing padding on the fly with help of "collate_fn" where I pass my padding function. I have faced the issue that when I am doing padding even 1 dim (just add one zero at the end of each object, then latent space is not trainable) while if I do absolutely the same in advance and save my features, afterward I will not use  "collate_fn" during the training then the latent space is trainable, what can be an issue?

**Comment by KevinMusgrave** [comment]
Can you use a custom class and print the dimensions of the data returned by the dataloader? 

```python
from pytorch_metric_learning.trainers import MetricLossOnly

class CustomMetricLossOnly(MetricLossOnly):
    def calculate_loss(self, curr_batch):
        data, labels = curr_batch

        # check the shape of data
        print("data", data.shape)

        embeddings = self.compute_embeddings(data)
        indices_tuple = self.maybe_mine_embeddings(embeddings, labels)
        self.losses["metric_loss"] = self.maybe_get_metric_loss(
            embeddings, labels, indices_tuple
        )
```

**Comment by tlscabinet** [comment]
Then should I call just "trainers.MetricLossOnly" after I have run class CustomMetricOnly?

**Comment by KevinMusgrave** [comment]
You should set your trainer to be ```CustomMetricLossOnly```

```python
trainer = CustomMetricLossOnly(...)
trainer.train(num_epochs=num_epochs)
```

**Comment by tlscabinet** [comment]
So the output if (batch_size, dim1, padded_dim2) (64, 80, 615) as it have to be

**Comment by tlscabinet** [comment]
Basically I have done easy test: I have took my padded in advance data and run trainer without using "collate_fn" function, the latent space was trainable (shape of each batch was 65, 80, 615). Afterwards I have used "collate_fn" simply cut last 15 dimension and  afterwards I have found that latent space was not trainable again (shape of each batch was 65, 80, 600).
So seems that it can be a problem somewhere in the the package, what do you think?

**Comment by KevinMusgrave** [comment]
Can you be specific about what you mean by "not trainable"?

**Comment by tlscabinet** [comment]
Sure, Trainable means that I am getting clusters in the embedding space, plus classification metrics is pretty high, while not trainable means that the UMAP embeddings looks like a noise plus classification metrics are low

**Comment by KevinMusgrave** [comment]
Are you using GlobalEmbeddingSpaceTester to view the UMAP embeddings and calculate accuracy?

**Comment by tlscabinet** [comment]
Yes I am

**Comment by KevinMusgrave** [comment]
Can you use this custom tester to see if the data shape and embedding shapes are correct?

```python
class CustomTester(GlobalEmbeddingSpaceTester):
    def get_embeddings_for_eval(self, trunk_model, embedder_model, input_imgs):
        print("data.shape", input_imgs.shape)

        input_imgs = c_f.to_device(
            input_imgs, device=self.data_device, dtype=self.dtype
        )
        trunk_output = trunk_model(input_imgs)
        if self.use_trunk_output:
            return trunk_output
        return embedder_model(trunk_output)


    def get_all_embeddings_for_all_splits(
        self,
        dataset_dict,
        trunk_model,
        embedder_model,
        splits_to_compute_embeddings,
        collate_fn,
    ):
        embeddings_and_labels = {}
        for split_name in splits_to_compute_embeddings:
            logging.info("Getting embeddings for the %s split" % split_name)
            embeddings_and_labels[split_name] = self.get_all_embeddings(
                dataset_dict[split_name], trunk_model, embedder_model, collate_fn
            )

        for k, (embeddings, labels) in embeddings_and_labels.items():
            print(k, embeddings.shape) 

        return embeddings_and_labels

```

**Comment by tlscabinet** [comment]
All what I have expected to see, I saw. Btw still not able to add collate_fn function into tester, maybe tomorrow I will also modify the init to be able to add

**Comment by KevinMusgrave** [comment]
What version of this library are you using?

```python
import pytorch_metric_learning
print(pytorch_metric_learning.__version__)
```

**Comment by tlscabinet** [comment]
0.9.97

**Comment by tlscabinet** [comment]
Hello, once more, I have done few more tests, still same issue, can you contact me privately, I will give more information.

**Comment by KevinMusgrave** [comment]
If you are using ```logging_presets.HookContainer.end_of_epoch_hook``` and you want the tester to use a custom ```collate_fn```, you must pass in ```collate_fn``` like this:

```python
import pytorch_metric_learning.utils.logging_presets as logging_presets
record_keeper, _, _ = logging_presets.get_record_keeper("example_logs", "example_tensorboard")
hooks = logging_presets.get_hook_container(record_keeper)
end_of_epoch_hook = hooks.end_of_epoch_hook(tester, 
                                            dataset_dict, 
                                            model_folder, 
                                            test_interval = 5,
                                            patience = None,
                                            test_collate_fn = collate_fn) ### this line
```

See the documentation for more info: https://kevinmusgrave.github.io/pytorch-metric-learning/logging_presets/#hookcontainer
